File: Nim/Types/objects.nim
type
  Animal* = object
    name*, species*: string
    age: int

proc sleep*(a: var Animal) =
  a.age += 1

proc dead*(a: Animal): bool =
  result = a.age > 20

var carl: Animal
carl = Animal(name : "Carl",
              species : "L. glama",
              age : 12)

let joe = Animal(name : "Joe",
                 species : "H. sapiens",
                 age : 23)

assert(not carl.dead)
for i in 0..10:
  carl.sleep()
assert carl.dead


# ref objects

let mittens: ref Animal = new(Animal)

mittens.name = "Mittens"
mittens.species = "P. leo"
mittens.age = 6

# or just create ref type of animal

type
  AnimalRef* = ref Animal

let spot = AnimalRef(
  name: "Spot",
  species: "C. lupus",
  age: 1)

type
  Thing* = ref object
    positionX*, positionY*: int

File: TS/Types/objects.ts
type Animal = {
  name: string
  species: string
  age: number
}

function sleep(a: Animal) {
  return a.age += 1
}

function dead(a: Animal): boolean {
  return a.age > 20
}

let carl: Animal = {
  name : "Joe",
  species : "H. sapiens",
  age : 23
}